### PR TITLE
adding filter feature selection methods for uplift modeling

### PR DESCRIPTION
**Pull request by zhenyuz0500**
Adding feature selection filter methods for uplift modeling.
Methods include: F filter, LR filter, KL filter, Chi filter, ED filter (more details to be described by a upcoming paper)

To use these methods:
from causalml.feature_selection import FilterSelect

An example notebook will be added later.

### CODE REVIEW

File: causalml/feature_selection/filters.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+class FilterSelect:
+    """A class for feature importance methods.
+    """
+
+    def __init__(self):
+        return
+
+    def _convert_treatment(self, data, T_name_before, treatment_key):
+        """
+        Convert text treatment labels into 0 (control) and 1 (treated).
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+        """
+        T_dummies = pd.get_dummies(data[T_name_before])
+        T_dummies_treatment = T_dummies[treatment_key]   
+        return T_dummies_treatment
+
+
+    def _filter_F_one_feature(self, data, T_name, X_name, y_name): 
+        """
+        Conduct F-test of the interaction between treatment and one feature.
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
+"""
+Filter feature selection methods for uplift modeling
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Perhaps move to main docstring.

File: causalml/feature_selection/filters.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+            'p_value': F_test.pvalue, 
+            'misc': 'df_num: {}, df_denom: {}'.format(F_test.df_num, F_test.df_denom), 
+        }, index=[0]).reset_index(drop=True)
+
+        return F_test_result
+
+
+    def filter_F(self, data, T_name, X_name_list, y_name):
+        """
+        Rank features based on the F-statistics of the interaction.
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+        all_result = all_result.sort_values(by='score', ascending=False)
+        all_result['rank'] = all_result['score'].rank(ascending=False)
+
+        return all_result
+
+
+    def _filter_LR_one_feature(self, data, T_name, X_name, y_name, disp=True): 
+        """
+        Conduct LR (Likelihood Ratio) test of the interaction between treatment and one feature.
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+            'p_value': LR_pvalue,
+            'misc': 'df: {}'.format(LR_df), 
+        }, index=[0]).reset_index(drop=True)
+
+        return LR_test_result
+
+
+    def filter_LR(self, data, T_name, X_name_list, y_name, disp=True):
+        """
+        Rank features based on the LRT-statistics of the interaction.
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+            Counts of conversions by treatment groups, of the form: 
+            {'control': {0: 10, 1: 8}, 'treatment1': {0: 5, 1: 15}}
+        nodeSummary: dict
+            Probability of conversion and group size by treatment groups, of 
+            the form:
+            {'control': [0.490, 500], 'treatment1': [0.584, 500]}
+
+        References
+        ----------
+        - results: upliftpy.UpliftTreeClassifier.group_uniqueCounts()
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Are these in the wrong section?

File: causalml/feature_selection/filters.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        treatment_group_keys = results_series.index.levels[0].tolist()
+        y_name_keys = results_series.index.levels[1].tolist()
+
+        results = {}
+        for ti in treatment_group_keys: 
+            results.update({ti: {}}) 
+            # results[ti] = {} # alternatively
+            for ci in y_name_keys:
+                results[ti].update({ci: results_series[ti, ci]}) 
+                # results[ti][ci] = results_series[ti, ci] # alternatively
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Consider removing.

File: causalml/feature_selection/filters.py [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+    def _kl_divergence(self, pk, qk):
+        """
+        Calculate KL Divergence for binary classification.
+
+        Parameters
+        ----------
+        pk, qk : float
+        """
+        # Formula
+        # S = sum(np.array(pk) * np.log(np.array(pk) / np.array(qk)))
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Consider removing.

File: causalml/feature_selection/filters.py [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+        pc = nodeSummary[control_group][0]
+        d_res = 0
+        for treatment_group in nodeSummary:
+            if treatment_group != control_group:
+                d_res += self._kl_divergence(nodeSummary[treatment_group][0], pc)
+        return d_res
+
+    def _evaluate_ED(self, nodeSummary, control_group='control'):
+        """
+        Calculate the multi-treatment unconditional D (one node)
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+        d_res = 0
+        for treatment_group in nodeSummary:
+            if treatment_group != control_group:
+                d_res += 2 * (nodeSummary[treatment_group][0] - pc)**2
+        return d_res
+
+    def _evaluate_Chi(self, nodeSummary, control_group='control'):
+        """
+        Calculate the multi-treatment unconditional D (one node)
+        with Chi-Square as split Evaluation function.
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+            'misc': 'number_of_bins: {}'.format(min(n_bins, x_bin.max()+1)),# format(n_bins),
+        }, index=[0]).reset_index(drop=True)
+
+        return(D_result)
+
+    def filter_D(self, data, X_name_list, y_name, 
+                 n_bins=10, method='KL', control_group='control',
+                 experiment_group_column='treatment_group_key'):
+        """
+        Rank features based on the chosen divergence measure.
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Could we add the args here?

File: causalml/feature_selection/filters.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+
+        all_result = all_result.sort_values(by='score', ascending=False)
+        all_result['rank'] = all_result['score'].rank(ascending=False)
+
+        return all_result
+
+
+    # Get node summary - a function 
+    def _GetNodeSummary(self, data, 
+                        experiment_group_column='treatment_group_key',
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Should we consider another default argument than 'treatment_group_key'?

File: causalml/feature_selection/filters.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+        # Counts of conversions by treatment group
+        results_series = data.groupby([experiment_group_column, y_name]).size()
+        
+        treatment_group_keys = results_series.index.levels[0].tolist()
+        y_name_keys = results_series.index.levels[1].tolist()
+
+        results = {}
+        for ti in treatment_group_keys: 
+            results.update({ti: {}}) 
+            # results[ti] = {} # alternatively
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Consider removing.

File: causalml/feature_selection/filters.py [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+        # Formula
+        # S = sum(np.array(pk) * np.log(np.array(pk) / np.array(qk)))
+        if qk < 0.1**6:
+            qk = 0.1**6
+        elif qk > 1 - 0.1**6:
+            qk = 1 - 0.1**6
+        S = pk * np.log(pk / qk) + (1-pk) * np.log((1-pk) / (1-qk))
+        return S
+
+    def _evaluate_KL(self, nodeSummary, control_group='control'):
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Same question as above regarding the default argument. I think we use 0 as the default control label elsewhere.

File: causalml/feature_selection/filters.py [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+                d_res += (
+                    (nodeSummary[treatment_group][0] - pc)**2 / max(0.1**6, pc) 
+                    + (nodeSummary[treatment_group][0] - pc)**2 / max(0.1**6, 1-pc)
+                )
+        return d_res
+
+
+    def _filter_D_one_feature(self, data, X_name, y_name, 
+                              n_bins=10, method='KL', control_group='control',
+                              experiment_group_column='treatment_group_key'):
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: As above, consider changing the default argument.

File: causalml/feature_selection/filters.py [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+        if method == 'KL':
+            evaluationFunction = self._evaluate_KL
+        elif method == 'ED':
+            evaluationFunction = self._evaluate_ED
+        elif method == 'Chi':
+            evaluationFunction = self._evaluate_Chi
+
+        totalSize = len(data.index)
+        x_bin = pd.qcut(data[X_name].values, n_bins, labels=False, 
+                        duplicates='drop')
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Do we want to silently drop the duplicates or eg add some jitter? Recall the problem with small sample sizes.

File: causalml/feature_selection/filters.py [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+            'score': d_res, 
+            'p_value': None,
+            'misc': 'number_of_bins: {}'.format(min(n_bins, x_bin.max()+1)),# format(n_bins),
+        }, index=[0]).reset_index(drop=True)
+
+        return(D_result)
+
+    def filter_D(self, data, X_name_list, y_name, 
+                 n_bins=10, method='KL', control_group='control',
+                 experiment_group_column='treatment_group_key'):
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: The default arg :)

File: causalml/feature_selection/filters.py [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+            )
+            all_result = pd.concat([all_result, one_result])
+
+        all_result = all_result.sort_values(by='score', ascending=False)
+        all_result['rank'] = all_result['score'].rank(ascending=False)
+
+        return all_result
+
+    def get_importance(self, data, X_name_list, y_name, method, 
+                      experiment_group_column='treatment_group_key',
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: Same as above regarding the default arg.

File: causalml/feature_selection/filters.py [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+                experiment_group_column=experiment_group_column, 
+            )
+            all_result = pd.concat([all_result, one_result])
+
+        all_result = all_result.sort_values(by='score', ascending=False)
+        all_result['rank'] = all_result['score'].rank(ascending=False)
+
+        return all_result
+
+    def get_importance(self, data, X_name_list, y_name, method,
```
Comment (t-tte) [inline causalml/feature_selection/filters.py]: More generally, I think it would make sense to use same conventions as elsewhere in the package, ie use "X", "treatment", "y" and "control_name" as the argument names.

File: causalml/feature_selection/filters.py (hunk 1 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: This is an intermediate function that is not user-facing. It seems sklearn skips the args for intermediate functions (with _ prefix), for example: https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/ensemble/_forest.py#L2172. But we can make our own rules. If we decide to add args for all functions, I will add args here.

File: causalml/feature_selection/filters.py (hunk 3 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: adding descriptions to .rst file?

File: causalml/feature_selection/filters.py (hunk 13 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: any suggestions?

File: causalml/feature_selection/filters.py (hunk 7 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: good catch, the references need to be updated.

File: causalml/feature_selection/filters.py (hunk 14 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: agree

File: causalml/feature_selection/filters.py (hunk 8 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: agree

File: causalml/feature_selection/filters.py (hunk 15 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: The current data generation function generates 'control' and 'treatment', and this setting would compile with the DGP well. However, this is an intermediate function, that is not user-facing.

File: causalml/feature_selection/filters.py (hunk 17 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: Good call. One difficulty is that the scale of X is unknown, therefore, adding 10**(-8) may not be always safe. One option is to leave it to users to manipulate X outside, the other option, is to add a value that is proportional to the scale of X, e.g. at the scale of 10**(-8)*median(X).

File: causalml/feature_selection/filters.py (hunk 20 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: hmm, one difference is that, in other functions, X and y are values in list format, but X_name_list and y_name is string here. The consideration is that feature importance will come with feature name as outputs for interpretation, therefore only values will be insufficient. Instead, we require inputs in (data, X_name_list, y_name) format.

File: causalml/feature_selection/filters.py (hunk 1 above)
Comment (t-tte) [reply causalml/feature_selection/filters.py]: I'm also fine with the convention of not specifying the args for non-public functions. I was purely thinking it from the point of view that it was not immediately clear to me what "T_name_before" means.

File: causalml/feature_selection/filters.py (hunk 13 above)
Comment (t-tte) [reply causalml/feature_selection/filters.py]: We're using 0 elsewhere in the package. Not sure that it matters that much if it is a non-public function.

File: causalml/feature_selection/filters.py (hunk 3 above)
Comment (t-tte) [reply causalml/feature_selection/filters.py]: NVM, I think it's fine as it is.

File: causalml/feature_selection/filters.py (hunk 20 above)
Comment (t-tte) [reply causalml/feature_selection/filters.py]: True. In terms of pre-existing convention, it looks like we're using 'features' for a list of feature names in the explainer class here: https://github.com/uber/causalml/blob/master/causalml/inference/meta/explainer.py For y_name, I'm not sure that any convention exists.

File: causalml/feature_selection/filters.py (hunk 17 above)
Comment (t-tte) [reply causalml/feature_selection/filters.py]: In the interest of explicitness, I think it may make sense to switch to duplicates='raise' and let the user manipulate X as needed.

File: causalml/feature_selection/filters.py (hunk 1 above)
Comment (jeongyoonlee) [reply causalml/feature_selection/filters.py]: I second to @t-tte. When it might not be straightforward to others, it's better to have docstrings.

File: causalml/feature_selection/filters.py (hunk 13 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: hmm, is 0 a default column name, or a default value within the experiment label column? To clarify, experiment_group_column is a column name, and data[experiment_group_column] will be the list of control / treatment labels.

File: causalml/feature_selection/filters.py (hunk 4 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: Updated "T_name" to a more interpretable name "treatment_indicator". This arg will only be used for internal functions. Args are added below.

File: causalml/feature_selection/filters.py (hunk 5 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: done

File: causalml/feature_selection/filters.py (hunk 6 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: done

File: causalml/feature_selection/filters.py (hunk 10 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: done

File: causalml/feature_selection/filters.py (hunk 11 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: done

File: causalml/feature_selection/filters.py (hunk 12 above)
Comment (zhenyuz0500) [reply causalml/feature_selection/filters.py]: done
